File: data-structures/src/test/java/vn/khanhpdt/playgrounds/datastructures/stacks/StackTest.java
package vn.khanhpdt.playgrounds.datastructures.stacks;

import org.junit.Test;
import vn.khanhpdt.playgrounds.datastructures.TestUtils;
import vn.khanhpdt.playgrounds.datastructures.nodes.SinglyLinkedNode;

import java.util.Collections;
import java.util.List;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;

/**
 * @author khanhpdt
 */
public class StackTest {

	@Test
	public void testPush() throws Exception {
		Stack<SinglyLinkedNode> stack = new Stack<>();

		SinglyLinkedNode newNode = SinglyLinkedNode.random();
		stack.push(newNode);

		assertThat(stack.peek(), is(newNode));
	}
	@Test
	public void testPop() throws Exception {
		Stack<SinglyLinkedNode> stack = new Stack<>();

		SinglyLinkedNode newNode = SinglyLinkedNode.random();
		stack.push(newNode);

		assertThat(stack.pop(), is(newNode));
	}

	@Test
	public void testPushAndPopMultipleItems() throws Exception {
		List<SinglyLinkedNode> nodes = TestUtils.randomSinglyNodes(10);

		Stack<SinglyLinkedNode> stack = new Stack<>();
		nodes.forEach(stack::push);

		Collections.reverse(nodes);
		nodes.forEach(n -> assertThat(stack.pop(), is(n)));
	}
}

File: golang/src/dsalg/ds/stack/stack_test.go
package stack

import "testing"

func TestPush(t *testing.T) {
	var s = new(Stack)
	s.Push("a")
	s.Push("b")

	var stackAsString = s.String()
	if stackAsString != "b -> a" {
		t.Errorf(`stack = %q, want: "b -> a"`, stackAsString)
	}
}

func TestPop(t *testing.T) {
	var s = new(Stack)
	s.Push("a")
	s.Push("b")
	s.Push("c")

	s.Pop()

	var stackAsString = s.String()
	if stackAsString != "b -> a" {
		t.Errorf(`stack = %q, want: "b -> a"`, stackAsString)
	}

	s.Pop()

	stackAsString = s.String()
	if stackAsString != "a" {
		t.Errorf(`stack = %q, want: "a"`, stackAsString)
	}

	s.Pop()

	stackAsString = s.String()
	if stackAsString != "" {
		t.Errorf(`stack = %q, want: ""`, stackAsString)
	}
}
